### PR TITLE
refactor(sql): optimize sql query parser

### PR DESCRIPTION
**Pull request by lilykuang**
### CATEGORY

Choose one

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [x] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

When I was researching on optimizing sql parser, I have tried using [`flatten`](https://sqlparse.readthedocs.io/en/latest/analyzing/#sqlparse.sql.TokenList.flatten) method to reduce processing time. However, `flatten` turns sql query into 1 dimensional iterator and it loses the ability to identify item as tokenList. TokenList is used as identifier to process as a table, see below https://github.com/apache/incubator-superset/blob/e5c3e8964d8933069569809325200474157592c2/superset/sql_parse.py#L171-L175

I have found that  `_extract_from_token` was recursively called on the same `item` for n times (n is length of `item.tokens`). It could be called only one time for `item`

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/sql_parse.py [hunk 1]
```diff
@@ -222,6 +220,15 @@ def __extract_from_token(self, token: Token):  # pylint: disable=too-many-branch
                     if not self.__is_identifier(token2):
                         self.__extract_from_token(item)
 
+    def __process_token_group(self, tokens):  # flatten nested tokens to 1D array
```
Comment (john-bodley) [inline superset/sql_parse.py:223]: `sqlparse` has a `flatten` method and thus you may want to explore whether this serves the same purpose as said function.

**Review by john-bodley — COMMENTED** [review]
Could you please and a summary in the PR description to describe how the parsing has been optimized?

**Comment by villebro** [comment]
While I don't doubt this does what the PR description says, it would be great if we could transfer some understanding that you've gathered during research for this PR into the code/docs. Reviewing this is quite difficult in part because of the following:
- Docstrings are very short or missing. Perhaps some examples could be added to the docstrings to explain what kind of tokens a typical SQL statement is decomposed into.
- Documentation for `sqlparse` is also quite incomprehensive.
- Variable names are not very descriptive. For example, `subtoken` might be more explanatory than `token2`.

This might be overkill, but it would be great if we could add a test that ensures that `self._extract_from_token` for a certain test case isn't called more than x times. That would make sure this parsing logic stays performant in the future, too.

**Review by villebro — APPROVED** [review]
LGTM, thanks for fixing!

**Pull request merged**